appinventor/blocklyeditor/src/block_svg.js

```diff
@@ -300,37 +373,11 @@ Blockly.BlockSvg.prototype.renderHere = function(opt_bubble) {
     if (this.collapsed_ == collapsed) {
       return;
     }
-    var renderList = [];
-    // Show/hide the inputs.
-    for (var x = 0, input; input = this.inputList[x]; x++) {
-      // No need to collect renderList if rendering down.
-      input.setVisible(!collapsed);
-    }
-
-    var COLLAPSED_INPUT_NAME = '_TEMP_COLLAPSED_INPUT';
-    if (collapsed) {
-      var icons = this.getIcons();
-      for (var i = 0; i < icons.length; i++) {
-        icons[i].setVisible(false);
-      }
-      var text = this.toString(Blockly.COLLAPSE_CHARS);
-      this.appendDummyInput(COLLAPSED_INPUT_NAME).appendField(text).init();
-    } else {
-      this.removeInput(COLLAPSED_INPUT_NAME);
-      // Clear any warnings inherited from enclosed blocks.
-      this.setWarningText(null);
-    }
-    Blockly.BlockSvg.superClass_.setCollapsed.call(this, collapsed);
-
-    if (!renderList.length) {
-      // No child blocks, just render this block.
-      renderList[0] = this;
-    }
-    if (this.rendered) {
```

Comment:
I'm concerned about the removal of this logic. In particular, we initialize blocks with this.rendered set to false because we don't want any render operations happening when we perform logic upgrades to the language. By removing this check, I think the blocks are being marked as rendered, and then loading blows up. To test, build this branch and then create a project with some blocks. Collapse the blocks and save the project. Reload App Inventor and you will get an error saying that the blocks workspace won't load. If you pause on caught exceptions in the debugger, you'll see that the bad block code is trying to do some operation even though the block isn't attached to the DOM yet.

Reply:
@ewpatton Thanks for the detailed repro steps!

The fix was super easy, I just needed to check if the block was rendered before calling render.

But writing a test for this has been much trickier (and taken a lot longer). This bug only affects component blocks, so I'm having to mock the component database. I'll give it another shot tomorrow, but any advice would be appreciated.